src/posit/connect/external/databricks.py

```diff
@@ -5,73 +5,109 @@
 
+    def sql_credentials_provider(self, *args, **kwargs):
+        """The sql connector attempts to call the credentials provider w/o any args.
 
-def viewer_credentials_provider(
-    client: Optional[Client] = None, user_session_token: Optional[str] = None
-) -> Optional[CredentialsProvider]:
-    # If the content is not running on Connect then viewer auth should
-    # fall back to the locally configured credentials hierarchy
-    if is_local():
-        return None
+        The SQL client's `ExternalAuthProvider` is not compatible w/ the SDK's implementation of
+        `CredentialsProvider`, so create a no-arg lambda that wraps the args defined by the real caller.
+        This way we can pass in a databricks `Config` object required by most of the SDK's `CredentialsProvider`
+        implementations from where `sql.connect` is called.
 
-    if client is None:
-        client = Client()
+        https://github.com/databricks/databricks-sql-python/issues/148#issuecomment-2271561365
+        """
+        return lambda: self.__call__(*args, **kwargs)
 
-    # If the user-session-token wasn't provided and we're running on Connect then we raise an exception.
-    # user_session_token is required to impersonate the viewer.
-    if user_session_token is None:
-        raise ValueError(
-            "The user-session-token is required for viewer authentication."
+    def auth_type(self) -> str:
+        """Returns the auth type currently in use.
+
+        The databricks-sdk client uses the configurated auth_type to create
+        a user-agent string which is used for attribution. We should only
+        overwrite the auth_type if we are using the PositCredentialsStrategy (non-local),
+        otherwise, we should return the auth_type of the configured local_strategy instead
+        to avoid breaking someone elses attribution.
+
+        https://github.com/databricks/databricks-sdk-py/blob/v0.29.0/databricks/sdk/config.py#L261-L269
+
+        NOTE: The databricks-sql client does not use auth_type to set the user-agent.
+        https://github.com/databricks/databricks-sql-python/blob/v3.3.0/src/databricks/sql/client.py#L214-L219
+        """
+        if _is_local():
+            return self.local_strategy.auth_type()
+        else:
+            return "posit-oauth-integration"
+
+    def __call__(self, *args, **kwargs) -> CredentialsProvider:
+        # If the content is not running on Connect then fall back to local_strategy
+        if _is_local():
+            return self.local_strategy(*args, **kwargs)
+
+        # If the user-session-token wasn't provided and we're running on Connect then we raise an exception.
+        # user_session_token is required to impersonate the viewer.
+        if self.user_session_token is None:
+            raise ValueError(
+                "The user-session-token is required for viewer authentication."
+            )
+
+        if self.client is None:
+            self.client = Client()
```

Comment:
Could you help me understand the situation in which self.client is None?  My concern here if someone needs to set the endpoint or api_key directory instead of using the environment variables.

Reply:
`self.client` might be None when the content is running locally in which case the user can't use the `PositCredentialsStrategy`.  We would fallback to the `local_strategy` instead. When the content is running on Connect then we know that Connect will set `CONNECT_API_KEY` and `CONNECT_SERVER_URL` for the content so the `Client()` would be allowed to initialize

Reply:
Got it. Thank you!